keras/src/layers/rnn/rnn.py

```diff
@@ -382,6 +382,20 @@ def call(
                 initial_state = self.get_initial_state(
                     batch_size=ops.shape(sequences)[0]
                 )
+
+        # Check for batch size mismatch when stateful
+        if self.stateful:
+            actual_batch_size = ops.shape(sequences)[0]
+            expected_batch_size = (self.states[0].shape[0] 
+                                  if self.states else None)
+            if (expected_batch_size is not None and 
+                not ops.equal(actual_batch_size, expected_batch_size)):
```

Comment:
You cannot do `ops.equal` in an `if`. `ops.equal` returns a tensor, this won't work in graph mode / jax.jit.